### ISSUE TITLE
GenServer `:tls_alert` error on handshake failure

### ISSUE THREAD

**Issue by danschultzer** [mtrudel/thousand_island#49]
`ThousandIsland.Handler` shuts down with an unexpected reason when there's a TLS handshake error, and reports this to the error logger:

```
17:01:55.000 [error] GenServer #PID<0.24370.1> terminating
** (stop) {:tls_alert, {:handshake_failure, 'TLS server: In state wait_finished received CLIENT ALERT: Fatal - Handshake Failure\n'}}
Last message: {:thousand_island_ready, %ThousandIsland.Socket{socket: {:sslsocket, {:gen_tcp, #Port<0.581>, :tls_connection, [option_tracker: #PID<0.24058.1>, session_tickets_tracker: :disabled, session_id_tracker: #PID<0.24059.1>]}, [#PID<0.24369.1>, #PID<0.24368.1>]}, transport_module: ThousandIsland.Transports.SSL, read_timeout: 60000, span: %ThousandIsland.Telemetry{span_name: :connection, telemetry_span_context: #Reference<0.4197377238.932970501.69923>, start_time: -576460732548559697, start_metadata: %{parent_telemetry_span_context: #Reference<0.4197377238.932970498.70998>, remote_address: {127, 0, 0, 1}, remote_port: 56792, telemetry_span_context: #Reference<0.4197377238.932970501.69923>}}}}
```

I think TLS handshake failure should be expected and is normal behavior in the client-server negotiation, and not sure it makes sense to shut it down this way. Also, `:ssl` already reports to the OTP logger. In my case I'm testing by accessing the server on a different hostname than what the cert is tied to, and I wouldn't expect anything in Bandit/Thousand Island to report abnormal behavior (I already know what went wrong client side).

I was thinking of adding a conditional in `ThousandIsland.Handler`:

```elixir
      def handle_info({:thousand_island_ready, socket}, {nil, state}) do
        ThousandIsland.Telemetry.span_event(socket.span, :ready)

        case ThousandIsland.Socket.handshake(socket) do
          {:ok, socket} -> {:noreply, {socket, state}, {:continue, :handle_connection}}
          {:error, {:tls_alert, error} = reason} -> {:stop, {:shutdown, reason}, {socket, state}}
          # Or if only the hanshake failure: {:error, {:tls_alert, {:handshake_failure, _} = reason} -> {:stop, {:shutdown, reason}, {socket, state}}
          {:error, reason} -> {:stop, reason, {socket, state}}
        end
      end
```

But maybe this is exposing too much of the underlying transport logic and it should be in `ThousandIsland.Transports.SSL`:

```elixir
def handshake(socket) do
  case :ssl.handshake(socket) do
    {:ok, socket} -> {:ok, socket}
    {:error, {:tls_alert, _error} = reason} -> {:error, {:shutdown, reason}}
    {:error, error} -> {:error, error}
  end
end
```

@mtrudel what do you think would be correct behavior here? I'm erring on the side of the GenServer closing down normally for any failed TLS negotiation, because the client would be aware of it in any case.

**Comment by mtrudel** [comment]
This is a tricky one. 

On the one hand I generally think that Thousand Island should express as little policy as possible, and this change would definitely result in us having more policy (we're determining what is and isn't an 'interesting' error on the behalf of the user). 

On the other hand, I sympathize with the argument that this particular error is often a noisy one. 

What about a 'raise_handshake_errors' config flag, defaulting to 'true'?

**Comment by danschultzer** [comment]
Yeah, that might be the solution but is it necessary to return an abnormal stop error? AFAIK the only thing it does is that it logs [an error](https://hexdocs.pm/elixir/1.12/GenServer.html#c:terminate/2) (you can see it in the [`:gen_server`](https://www.erlang.org/doc/man/gen_server.html#stop-3) source [here](https://github.com/erlang/otp/blob/1b3c6214d4bf359f7e5c143bef5c5ad9c90c5536/lib/stdlib/src/gen_server.erl#L1319-L1342)).

Couldn't it just return `{:stop, {:shutdown, reason}}` for all handshake errors instead? Is there any abnormal behavior that requires it to be logged here?

**Comment by danschultzer** [comment]
I'm just thining out loud here when you would need this error logged. Is this something that's best handled at low level with the GenServer, or should it instead be dealt with in `ThousandIsland` or at a higher level in any library using the `ThousandIsland.Handler` GenServer.

**Comment by danschultzer** [comment]
Actually with how `ThousandIsland.Handler` is set up I'm wondering if it's not really a concern here, and instead should be handled at a higher level. But the handshake is abstracted away so no way to hook in without making assumptions for how it works in the underlying layer. It is a tricky one (I guess the error could always be wrapped in a touple specifying what stage).

I can open a PR for the `raise_handshake_errors` config, but it feels a little off. What if I want to log some handshake errors but not others?

**Comment by mtrudel** [comment]
I just pushed up a PR that handles this quite nicely I think. 

I realized that so long as well continued to ensure that we called handle_error with the underlying tls error (thus giving the user a place to manage the error), we don't need to worry too much about the logging issues. This is really the exact same flow that happens today with timeouts and we don't log in that case, so there's prior art in the library already for this approach.

WDYT?
